dLux/apertures.py

```diff
@@ -0,0 +1,1430 @@
+from typing import TypeVar
+from dLux.utils import (get_positions_vector, get_pixel_positions)
+from abc import ABC, abstractmethod 
+import equinox as eqx
+import jax.numpy as np
+import jax 
+import functools
+
+jax.config.update("jax_enable_x64", True)
+
+Array = TypeVar("Array")
+Layer = TypeVar("Layer")
+Tensor = TypeVar("Tensor")
+Matrix = TypeVar("Matrix")
+Vector = TypeVar("Vector")
+
+
+__all__ = ["Aperture", "CompoundAperture", "SoftEdgedAperture", 
+    "SquareAperture", "SoftEdgedSquareAperture", "RectangularAperture",
+    "SoftEdgedRectangularAperture", "CircularAperture", 
+    "SoftEdgedCircularAperture", "AnnularAperture", 
+    "SoftEdgedAnnularAperture"]
+
+
+def cartesian_to_polar(coordinates : Tensor) -> Tensor:
+    """
+    Change the coordinate system from rectilinear to curvilinear.
+    
+    Parameters
+    ----------
+    coordinates : Tensor
+        The rectilinear coordinates.
+
+    Returns
+    -------
+    coordinates : Tensor
+        The curvilinear coordinates.
+    """
+    rho = np.hypot(coordinates[0], coordinates[1])
+    theta = np.arctan2(-coordinates[1], coordinates[0])
+    return np.array([rho, theta])
+
+
+class Aperture(eqx.Module, ABC):
+    """
+    An abstract class that defines the structure of all the concrete
+    apertures. An aperture is represented by an array, usually in the
+    range of 0. to 1.. Values in between can be used to represent 
+    soft edged apertures and intermediate surfaces. 
+
+    Attributes
+    ----------
+    pixels : int
+        The number of pixels along one edge of the array which 
+        represents the aperture.
+    x_offset : float, meters
+        The x coordinate of the centre of the aperture.
+    y_offset : float, meters
+        The y coordinate of the centre of the aperture.
+    theta : float, radians
+        The angle of rotation from the positive x-axis. 
+    phi : float, radians
+        The rotation of the y-axis away from the vertical and torward 
+        the negative x-axis. 
+    magnification : float
+        The radius of the aperture. The radius belongs to the smallest
+        circle that completely contains the aperture. For the math
+        nerds the infimum of the set of circles whose union with the
+        aperture is the aperture. 
+    pixel_scale : float, meters per pixel 
+        The length along one side of a square pixel. 
+    """
+    pixels : int
+    x_offset : float
+    y_offset : float
+    theta : float
+    phi : float
+    magnification : float
+    pixel_scale : float # Not gradable
```

Comment:
Having looked at the code, this technically can be graded (unless I missed something). I understand why we wouldn't want to grad it, but for example if there is some uncertainty in the mask we want to be able to HMC over it, so we can understand how it interacts with other parameters.

Furthermore, I don't actually think we want to actually define the pixel scale here. If everything is defined in physical units (which I think it should be, and already is in this code), then we can simply take the coordinate arrays from the wavefront object. I think this is important because otherwise we have two *different* pixel scales defined at the same time. Check out the `CompoundAperture` class [here](https://github.com/LouisDesdoigts/dLux/blob/main/dLux/layers.py) that uses the Wavefront class pixel scale.